go/arrow/extensions/json.go

```diff
@@ -0,0 +1,139 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extensions
+
+import (
+	"fmt"
+	"reflect"
+	"slices"
+
+	"github.com/apache/arrow/go/v18/arrow"
+	"github.com/apache/arrow/go/v18/arrow/array"
+	"github.com/apache/arrow/go/v18/internal/json"
+	"github.com/apache/arrow/go/v18/parquet/schema"
+)
+
+var jsonSupportedStorageTypes = []arrow.DataType{
+	arrow.BinaryTypes.String,
+	arrow.BinaryTypes.LargeString,
+	arrow.BinaryTypes.StringView,
+}
+
+// JSONType represents a UTF-8 encoded JSON string as specified in RFC8259.
+type JSONType struct {
+	arrow.ExtensionBase
+}
+
+// ParquetLogicalType implements pqarrow.ExtensionCustomParquetType.
+func (b *JSONType) ParquetLogicalType() schema.LogicalType {
+	return schema.JSONLogicalType{}
+}
+
+// NewJSONType creates a new JSONType with the specified storage type.
+// storageType must be one of String, LargeString, StringView.
+func NewJSONType(storageType arrow.DataType) (*JSONType, error) {
+	if !slices.Contains(jsonSupportedStorageTypes, storageType) {
+		return nil, fmt.Errorf("unsupported storage type for JSON extension type: %s", storageType)
+	}
+	return &JSONType{ExtensionBase: arrow.ExtensionBase{Storage: storageType}}, nil
+}
+
+func (b *JSONType) ArrayType() reflect.Type { return reflect.TypeOf(JSONArray{}) }
+
+func (b *JSONType) Deserialize(storageType arrow.DataType, data string) (arrow.ExtensionType, error) {
+	if !(data == "" || data == "{}") {
+		return nil, fmt.Errorf("serialized metadata for JSON extension type must be '' or '{}', found: %s", data)
+	}
+	return NewJSONType(storageType)
+}
+
+func (b *JSONType) ExtensionEquals(other arrow.ExtensionType) bool {
+	return b.ExtensionName() == other.ExtensionName() && arrow.TypeEqual(b.Storage, other.StorageType())
+}
+
+func (b *JSONType) ExtensionName() string { return "arrow.json" }
+
+func (b *JSONType) Serialize() string { return "" }
+
+func (b *JSONType) String() string {
+	return fmt.Sprintf("extension<%s[storage_type=%s]>", b.ExtensionName(), b.Storage)
+}
+
+// JSONArray is logically an array of UTF-8 encoded JSON strings.
+// Its values are unmarshaled to native Go values.
+type JSONArray struct {
+	array.ExtensionArrayBase
+}
+
+func (a *JSONArray) String() string {
+	b, err := a.MarshalJSON()
+	if err != nil {
+		panic(fmt.Sprintf("failed marshal JSONArray: %s", err))
+	}
+
+	return string(b)
+}
+
+func (a *JSONArray) Value(i int) any {
+	val := a.value(i)
+
+	var res any
+	if err := json.Unmarshal(val, &res); err != nil {
+		panic(err)
+	}
+
+	return res
+}
+
+func (a *JSONArray) ValueStr(i int) string {
+	return string(a.value(i))
+}
+
+func (a *JSONArray) value(i int) []byte {
+	val := a.ValueJSON(i)
+	b, err := val.MarshalJSON()
```

Comment:
Does this mean that taking the string value of a JSONArray's element first unmarshals the JSON, then marshals it again? Isn't it wasteful (and also potentially loses non-significant whitespace or other details)?

Reply:
The methods unfortunately obscure what actually happening, but since we're operating on `json.RawMessage` (an alias for `[]byte`) which assumes the wrapped data is already valid JSON, these calls are fairly minimal.

So `valueJSON` performs a null check against the validity bitmap and then a type conversion (`string` -> `[]byte`) on the array element if not null.

The implementation of `MarshalJSON` for the `json.RawMessage` type actually just returns the RawMessage bytes themselves, since it's assumed to already be valid JSON. The only additional logic it provides is to return `[]byte("null")` if the value is `nil`, so this is just to reuse trivial logic from the standard library rather than write it ourselves.

So in the typical case, this is a type conversion to `[]byte` and then the slice is passed back unchanged. No mutation of the encoded value itself.

Reply:
[source](https://cs.opensource.google/go/go/+/refs/tags/go1.23.0:src/encoding/json/stream.go;l=262)

Reply:
Ahah, ok. Perhaps you could add a quick comment explaining this (especially the `nil` conversion part), or perhaps this is obvious already for a Go developer?

Reply:
Sure! I've added some comments clarifying this.